### PR TITLE
add NaclBox interface

### PR DESCRIPTION
**Pull request by mikhailUshakoff**

### CODE REVIEW

File: NaclBox/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+## Description
+
+NaclBox perform naclbox encryption and decryption. Debot should get the signing box from the SigningBox interface and use it for encryption. For encryption, the browser must deduce curve256 keys from the signing box secret key.
+
+## Functions
+
+Since DeBot is a smart contract then all functions work asyncronously by design. It means that they don't return anything despite the fact that all have `returns` value in their specification. Result callback will be called later and it must have arguments defined in `returns` section.
+
+`getPublicKey` - get curve256 public key from signing box handle.
```
Comment (sertony) [inline NaclBox/README.md:18]: Maybe we are to rename it to make it more clear (to identify this public key is supposed to be used for the encryption)? e.g. `getEncryptPublicKey`
```suggestion
`getEncryptPublicKey` - get curve256 public key from signing box handle.
```

File: NaclBox/README.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+arguments: 
+
+	answerId: uint32 - function id of result callback
+        publicKey: uint256 - handle to signing box
```
Comment (sertony) [inline NaclBox/README.md]: Or, rename this parameter instead, e.g. `signPublicKey`

File: NaclBox/README.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+`encrypt` - Public key authenticated encryption. Encrypt and authenticate a message using the senders secret key, the receivers public key, and a nonce.
+
+arguments: 
+
+	answerId: uint32 - function id of result callback
+    	decrypted: bytes - data that must be encrypted encoded
+    	nonce: bytes - nonce	
+        their_public: uint256 - receiver's public key
+        publicKey: uint256 - handle to signing box
```
Comment (sertony) [inline NaclBox/README.md]: Maybe like this?
```
        theirEncryptPublicKey: uint256 - receiver's public key
        signPublicKey: uint256 - handle to signing box
```

File: NaclBox/README.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+`decrypt` - Decrypt and verify the cipher text using the receivers secret key, the senders public key, and the nonce.
+
+arguments: 
+
+	answerId: uint32 - function id of result callback
+    	encrypted: bytes - data that must be decrypted
+    	nonce: bytes - nonce	
+        their_public: uint256 - receiver's public key
+        publicKey: uint256 - handle to signing box
```
Comment (sertony) [inline NaclBox/README.md]: Same concern about naming, i.e.
```
        theirEncryptPublicKey: uint256 - receiver's public key
        signPublicKey: uint256 - handle to signing box
```

File: NaclBox/README.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+	decrypted: bytes - decrypted data
+
+
+
+## Declaration in Solidity
+
+```jsx
+interface INaclBox {
+	function getPublicKey(uint32 answerId, uint256 publicKey) external returns (uint256 publicKey);
```
Comment (sertony) [inline NaclBox/README.md]:
```suggestion
	function getEncryptPublicKey(uint32 answerId, uint256 signPublicKey) external returns (uint256 encryptPublicKey);
```

File: NaclBox/README.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	function encrypt(uint32 answerId, bytes decrypted, bytes nonce, uint256 their_public, uint256 publicKey) external returns (bytes encrypted, uint256 publicKey);
```
Comment (sertony) [inline NaclBox/README.md]:
```suggestion
	function encrypt(uint32 answerId, bytes decrypted, bytes nonce, uint256 theirEncryptPublicKey, uint256 signPublicKey) external returns (bytes encrypted, uint256 encryptPublicKey);
```

File: NaclBox/README.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	function decrypt(uint32 answerId, bytes encrypted, bytes nonce, uint256 their_public, uint256 publicKey) external returns (bytes decrypted);
```
Comment (sertony) [inline NaclBox/README.md]:
```suggestion
	function decrypt(uint32 answerId, bytes encrypted, bytes nonce, uint256 theirEncryptPublicKey, uint256 signPublicKey) external returns (bytes decrypted);
```

File: NaclBox/README.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+	answerId: uint32 - function id of result callback
+    	decrypted: bytes - data that must be encrypted encoded
+    	nonce: bytes - nonce	
+        their_public: uint256 - receiver's public key
+        publicKey: uint256 - handle to signing box
+
+returns: 
+
+	encrypted: bytes - encrypted data
+	pubkey: uint256 - curve256 public key deduce from signing box keys
```
Comment (sertony) [inline NaclBox/README.md]:
```
	encryptPublicKey: uint256 - curve256 public key deduce from signing box keys
```

File: NaclBox/README.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+`getPublicKey` - get curve256 public key from signing box handle.
+
+arguments: 
+
+	answerId: uint32 - function id of result callback
+        publicKey: uint256 - handle to signing box
+
+returns: 
+
+	pubkey: uint256 - curve256 public key deduce from signing box keys
```
Comment (sertony) [inline NaclBox/README.md]:
```
	encryptPublicKey: uint256 - curve256 public key deduce from signing box keys
```

**Review by sertony — CHANGES_REQUESTED** [review]

**Review by sertony — APPROVED** [review]

File: NaclBox/NaclBox.sol [hunk 10]
```diff
@@ -0,0 +1,34 @@
+pragma solidity >= 0.6.0;
+interface INaclBox {
+function getEncryptPublicKey(uint32 answerId, uint32 edHandle) external returns (uint256 encryptPublicKey);
+function encrypt(uint32 answerId, bytes decrypted, bytes nonce, uint256 theirEncryptPublicKey, uint32 edHandle) external returns (bytes encrypted, uint256 encryptPublicKey);
+function decrypt(uint32 answerId, bytes encrypted, bytes nonce, uint256 theirEncryptPublicKey, uint32 edHandle) external returns (bytes decrypted);
+}
+
+library NaclBox {
+
+	uint256 constant DEBOT_ADDR = da6e573c0282785d9f8e88a81e0f2665811148cb3bd9b5dcce6a079ab7009958;
```
Comment (Keshoid) [inline NaclBox/NaclBox.sol]: Can you rename it to `ID`? it is better to use `ID` in `getRequiredInterfaces` like `[... , NaclBox.ID]`

**Review by Keshoid — CHANGES_REQUESTED** [review]
Please, make example DeBot from example contract.

File: NaclBox/NaclBox.sol [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+function encrypt(uint32 answerId, bytes decrypted, bytes nonce, uint256 theirEncryptPublicKey, uint32 edHandle) external returns (bytes encrypted, uint256 encryptPublicKey);
+function decrypt(uint32 answerId, bytes encrypted, bytes nonce, uint256 theirEncryptPublicKey, uint32 edHandle) external returns (bytes decrypted);
+}
+
+library NaclBox {
+
+	uint256 constant ID = da6e573c0282785d9f8e88a81e0f2665811148cb3bd9b5dcce6a079ab7009958;
+	int8 constant DEBOT_WC = -31;
+
+	function getPublicKey(uint32 answerId, uint32 edHandle) public pure {
```
Comment (sertony) [inline NaclBox/NaclBox.sol:13]:
```suggestion
	function getEncryptPublicKey(uint32 answerId, uint32 edHandle) public pure {
```

**Review by sertony — CHANGES_REQUESTED** [review]
I also have some doubts about a possibility to derive the encryption public key (Nacl) for any signing box handle

**Comment by sertony** [comment]
I guess that it's better to work with EncryptionBox handle instead of SigningBox handles
https://www.notion.so/tonlabs/Crypto-Boxes-in-Core-79c05b38d231482a91ace893d8701747#ace7d561f83943c2a66f49ef40b92a7b

But the problem is that we don't have Encryption Boxes in SDK yet =\

@melsomino FYI

Besides it looks like it's better to convert this interface into EncryptionBoxInput interface (like SigningBoxInput)

**Pull request closed without merging**